packages/core/src/utils/security/handlers.ts

```diff
@@ -0,0 +1,130 @@
+
+type Headers = Dictionary<string, string>;
+
+export type AuthErr = { name: string; message: string; status: number; headers: Headers };
+export type SecurityScheme = { type: string; name: string; in?: string; scheme?: string };
+
+const genUnauthorisedErr = (msg: string): AuthErr => ({
+  name: 'Unauthorised',
+  message: 'Invalid security scheme used',
+  status: 401,
+  headers: { 'WWW-Authenticate': msg },
+});
+
+function isBasicToken(token: string) {
+  const tokenParts = Buffer.from(token, 'base64')
+    .toString()
+    .split(':');
+
+  return tokenParts.length === 2;
+}
+
+const httpBasic = {
+  test: ({ scheme, type }: SecurityScheme) => scheme === 'basic' && type === 'http',
+  handle: <R, I>(someInput: I, name: string, resource?: R) => {
+    const authorizationHeader = get(someInput, ['headers', 'authorization'], '');
+
+    return authorizationHeader
+      ? checkHeader<R>(authorizationHeader, resource)
+      : Either.left(genUnauthorisedErr('Basic realm="*"'));
+  },
+};
+
+function checkHeader<R>(authorizationHeader: string, resource?: R) {
+  const [schema, token] = authorizationHeader.split(' ');
+
+  const isBasicTokenGiven = token && isBasicToken(token);
+  const isBasicSchema = schema === 'Basic';
+
+  const handler = [
+    {
+      test: () => isBasicSchema && isBasicTokenGiven,
+      handle: () => Either.right(resource),
+    },
+    {
+      test: () => isBasicSchema,
+      handle: () => Either.left({ name: 'Forbidden', status: 403, message: 'Invalid credentials used', headers: {} }),
+    },
+  ].find(possibleHandler => possibleHandler.test());
+
+  return handler ? handler.handle() : Either.left(genUnauthorisedErr('Basic realm="*"'));
+}
+
+const apiKeyInHeader = {
+  test: ({ type, in: where }: SecurityScheme) => where === 'header' && type === 'apiKey',
+  handle: <R, I>(someInput: I, name: string, resource?: R) => {
+    const isAPIKeyProvided = get(someInput, ['headers', name.toLowerCase()]);
+
+    return when<R>(isAPIKeyProvided, name, resource);
+  },
+};
+
+const apiKeyInQuery = {
+  test: ({ type, in: where }: SecurityScheme) => where === 'query' && type === 'apiKey',
+  handle: <R, I>(someInput: I, name: string, resource?: R) => {
+    const isApiKeyInQuery = get(someInput, ['url', 'query', name]);
+
+    return when<R>(isApiKeyInQuery, name, resource);
+  },
+};
+
+const bearerHandler = <R, I>(msg: string, someInput: I, name: string, resource?: R) => {
+  return when<R>(isBearerToken(get(someInput, 'headers')), msg, resource);
+};
+
+const openIdConnect = {
+  test: ({ type }: SecurityScheme) => type === 'openIdConnect',
+  handle: bearerHandler.bind({}, 'OpenID'),
+};
+
+const bearer = {
+  test: ({ type, scheme }: SecurityScheme) => scheme === 'bearer' && type === 'http',
+  handle: bearerHandler.bind({}, 'Bearer'),
+};
+
+const oauth2 = {
+  test: ({ type }: SecurityScheme) => type === 'oauth2',
+  handle: bearerHandler.bind({}, 'OAuth2'),
+};
+
+const apiKeyInCookie = {
```

Comment:
There are lots of different ways to sort methods around, some folks alphabetize everything, or group public and private then alpha those, or they just group similar stuff together. 

I like to group and alphabetize within the group, but whatever approach you decide on, this method seems to a long way from its friends. :)